cmake/MacroGuardTest.cmake

```
@@ -0,0 +1,107 @@
+# PROJECT_NAME: bsoncxx or mongocxx.
+# GUARDED_MACROS: list of macros that should be guarded by prelude/postlude headers.
+# INCLUDE_FILTERS: list of regex filters for headers to be added to the tests (must be relative to PROJECT_SOURCE_DIR).
+# EXCLUDE_FILTERS: list of regex filters for headers to be excluded (applied after INCLUDE_FILTERS).
+# PROJECT_TEST_PROPERTIES: the (bsoncxx|mongocxx)_test_properties target to be linked with.
+function(add_macro_guard_test)
+    set(opt_args "")
+    set(single_args "PROJECT_NAME;PROJECT_TEST_PROPERTIES")
+    set(multi_args "GUARDED_MACROS;INCLUDE_FILTERS;EXCLUDE_FILTERS")
+
+    cmake_parse_arguments(PARSED "${opt_args}" "${single_args}" "${multi_args}" ${ARGN})
+
+    if(NOT "${PARSED_UNPARSED_ARGUMENTS}" STREQUAL "")
+        message(FATAL_ERROR "unrecognized argument: ${PARSED_UNPARSED_ARGUMENTS}")
+    endif()
+
+    foreach(required_arg PROJECT_NAME PROJECT_TEST_PROPERTIES GUARDED_MACROS INCLUDE_FILTERS)
+        if("${PARSED_${required_arg}}" STREQUAL "")
+            message(FATAL_ERROR "missing required argument: ${required_arg}")
+        endif()
```

```suggestion
        if(NOT DEFINED "PARSED_${required_arg}")
            message(FATAL_ERROR "missing required argument: ${required_arg}")
        endif()
```

This permits empty strings as values, but that may be part of the check. Alternatively, tweak the error message:

```suggestion
        if("${PARSED_${required_arg}}" STREQUAL "")
            message(FATAL_ERROR "Required argument ${required_arg} was not provided or was provided with an empty value")
        endif()
```